experimental/SetPartitions/src/LinearSetPartition.jl

```diff
@@ -0,0 +1,186 @@
+"""
+    LinearSetPartition
+
+LinearSetPartition represents a linear combination of set-partitions. See Chapter 5 in [Gro20](@cite).
+"""
+struct LinearSetPartition{S <: AbstractPartition, T <: RingElement}
+    coefficients :: Dict{S, T}
+
+    function LinearSetPartition{S, T}(coeffs::Dict{S, T}) where {S <: AbstractPartition, T <: RingElement}
+        key_types = Set(typeof(k) for k in keys(coeffs))
+        @req length(key_types) <= 1 "keys in the coefficients dictionary must be of the same subtype of AbstractPartition"
```

Comment:
```suggestion
        @req isconcretetype(S) "Linear combinations are only defined for concrete subtypes of AbstractPartition"
```
I think this is what you actually want here, as `S` may otherwise be `AbstractPartition` even though all keys in the dictionary are of the same concrete type